### ISSUE TITLE
Ideation and Roadmap

### ISSUE THREAD

**Issue by murrellb** [MurrellGroup/ConditionalDensityLayers.jl#1]
This package will house layers (usable in the Flux ecosystem) that map an embedding vector $S_i$ to a continuous probability distribution $D_i = P(X|S_i)$. For our purposes, we need to be able to evaluate $P(X=x_i|S_i)$ pointwise for a specific $x_i$ (and this must be with a very low computational cost), and to sample $x \sim D_i$ (and it is ok if this is a bit more expensive).

Strategies:
- Gaussian mixture model, where the network outputs the centroids, covariances (isotropic?), and weights.
  - Evaluating $P(X=x_i|S_i)$ is proportional to the number of gaussian components
  - $x \sim D_i$ is very cheap.
- Direct (but unnormalized) log-density output, analogous to discrete logits
  - Evaluating $P(X=x_i|S_i)$ needs the normalizing constant to be calculated, or at least approximated
  - $x \sim D_i$ can be obtained by eg. MCMC

Points to keep in mind:
- Sometimes $x$ will be bounded (eg. $-\pi$ to $\pi$) and sometimes unbounded. We will want to handle both cases.
- Need to design a common interface.

Related to look at:
- https://juliapackages.com/p/bijectors
- https://github.com/TuringLang/NormalizingFlows.jl
- https://docs.juliahub.com/ADCME/b8Ld2/0.5.2/flow/
